# K fold cross validation

## Context

One of the most common diseases in the world, diabetes is becoming more and more common among sufferers. Although the exact aetiology of diabetes is still unknown, researchers think that environmental lifestyle factors and genetics both contribute significantly to the disease.

Diabetes increases the chance of acquiring certain secondary health problems, including heart disease and nerve damage. Therefore, diabetes can be detected and treated early to avoid complications and lower the chance of developing serious health issues.
Despite being incurable, medicines and treatment can help manage it.

In order to better understand this disease in women, researchers at the Bio-Solutions lab intend to detect patients who are at risk of diabetes by using machine learning models.

To learn how to do cross validation on the train set and how to interpret the results, we will utilise the PIMA Indian Diabetes Dataset.


## Data Dictionary

* Pregnancies: Number of times pregnant
* Glucose: Plasma glucose concentration over 2 hours in an oral glucose tolerance test
* BloodPressure: Diastolic blood pressure (mm Hg)
* SkinThickness: Triceps skinfold thickness (mm)
* Insulin: 2-Hour serum insulin (mu U/ml)
* BMI: Body mass index (weight in kg/(height in m)^2)
* Pedigree: Diabetes pedigree function - A function that scores likelihood of diabetes based on family history.
* Age: Age in years
* Class: Class variable (0: the person is not diabetic or 1: the person is diabetic)



## Import libraries

In [ ]:
# to work with dataframes
import pandas as pd
import numpy as np

# to split data into train and test
from sklearn.model_selection import train_test_split

# to build logstic regression model
from sklearn.linear_model import LogisticRegression

# to create k folds of data and get cross validation score
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

# to ignore warnings
import warnings
warnings.filterwarnings('ignore')

## Load and view dataset

In [ ]:
df = pd.read_csv('pima-indians-diabetes.csv')

df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   Pedigree  Age  Class  
0     0.627   50      1  
1     0.351   31      0  
2     0.672   32      1  
3     0.167   21      0  
4     2.288   33      1

In [ ]:
# separating data into X and Y
X = df.drop(['Class'], axis = 1)
Y = df['Class']

In [ ]:
# creating train and test sets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.30, random_state=1, stratify = Y)

In [ ]:
# defining kfold
kfold = KFold(n_splits=10, random_state=1, shuffle = True)

# number of splits = 10

In [ ]:
# defining the model
model = LogisticRegression(random_state = 1)

# storing accuracy values of model for every fold in "results"
results = cross_val_score(model, X_train, Y_train, cv=kfold, scoring='accuracy')

In [ ]:
# let's see the value of accuracy for every fold
print(results)

[0.77777778 0.81481481 0.81481481 0.83333333 0.72222222 0.85185185
 0.74074074 0.66037736 0.77358491 0.79245283]


In [ ]:
# let's see the mean accuracy score
print("Accuracy: %.3f%% (%.3f%%)" % (results.mean()*100.0, results.std()*100.0))

Accuracy: 77.820% (5.458%)


## Interpretation of scores
- Mean accuracy is 77.82% with a standard deviation of 5.46%
- So, we can say that the accuracy on any unseen data will lie between 72.4 (mean - standard deviation) and 83.3 (mean + standard deviation) with a confidence of 67%.